xtest/PLAN.md

```diff
@@ -0,0 +1,301 @@
+- Sequential test execution without parallelization
+- Extensive SDK version combinations creating massive parameterization
+- Module-scoped fixtures that cannot be shared efficiently
+- No test categorization or selective execution strategy
+
+---
+
+## 1. Test Execution Parallelization ⚡
+
+### 1.1 Add pytest-xdist for Parallel Execution
+**Impact: HIGH (50-75% time reduction)**
+- Currently NOT installed (missing from requirements.txt)
+- Can run tests in parallel across CPU cores
+- Implementation:
+  ```bash
+  pip install pytest-xdist
+  pytest -n auto  # Auto-detect CPU cores
+  pytest -n 8     # Explicit worker count
+  ```
+
+### 1.2 Ensure Thread-Safety
+**Impact: MEDIUM (prerequisite for parallelization)**
+- Global `cipherTexts` dictionaries in test_tdfs.py:14 need locking
+- Temporary namespace creation (conftest.py:195-199) needs isolation
+- File I/O operations need unique paths per worker
+
+### 1.3 Use pytest-xdist Scopes
+**Impact: MEDIUM**
+- Group tests by module/class to share fixtures
+- `--dist loadscope` for better fixture reuse
+- `--dist loadfile` for file-level isolation
+
+---
+
+## 2. Fixture Optimization 🔧
+
+### 2.1 Analyze Fixture Scope Efficiency
+**Current State:**
+- 29 fixtures with `scope="module"` (conftest.py)
+- 5 fixtures with `scope="session"`
+- Most expensive fixtures (namespace creation, attributes) are module-scoped
+
+**Improvements:**
+- Promote expensive setup fixtures to `session` scope where possible
+- Examples:
+  - `temporary_namespace` (line 194) → session scope with cleanup
+  - `temporary_attribute_*` fixtures → session scope
+  - `pt_file` generation (line 156) → already optimal
+
+### 2.2 Lazy Fixture Loading
+**Impact: MEDIUM**
+- Implement lazy evaluation for fixtures only used in subset of tests
+- Use `pytest.mark.usefixtures` selectively
+- Defer SDK version loading until needed
+
+### 2.3 Fixture Caching Enhancement
+**Impact: LOW-MEDIUM**
+- Current `cipherTexts` cache (test_tdfs.py:14) is good but limited
+- Expand to cache:
+  - Attribute definitions
+  - KAS configurations
+  - Policy evaluations
+
+---
+
+## 3. Test Organization & Selective Execution 🎯
+
+### 3.1 Implement Test Markers
+**Impact: HIGH (enables targeted runs)**
+- Add markers for test categories:
+  ```python
+  @pytest.mark.fast        # < 1 second
+  @pytest.mark.slow        # > 5 seconds
+  @pytest.mark.encryption  # Tests encrypt operations
+  @pytest.mark.decryption  # Tests decrypt operations
+  @pytest.mark.abac        # Policy tests
+  @pytest.mark.integration # Full SDK integration
+  @pytest.mark.unit        # Isolated logic
+  ```
+
+### 3.2 SDK Version Matrix Reduction
+**Impact: HIGH (reduce test count from 696)**
+- Current: All SDK combinations (go@v0.24.0, go@main, js@v0.4.0, js@main, java variants)
+- Strategy:
+  - **Smoke tests**: One version combination per SDK (reduce 80% of tests)
+  - **Compatibility matrix**: Weekly/nightly runs for full cross-version
```

Comment:
Optional addition: 'full regression' mode, perhaps runs on release branches, perhaps only on release branches of opentdf/service?